### DISCUSSION TITLE
How can I connect Waveshare 0.96 OLED via I2C to RPI Pico and program it using micropython?

### DISCUSSION THREAD

**Discussion by mlajciak** [micropython/micropython#13275] [Hardware & Peripherals]
Hi,

I need to connect one of these:
https://www.waveshare.com/wiki/0.96inch_OLED_(B)#Introduction
to a RPI Pico. I only have I2C left for this so I need to use it. I resolder the BS1 resistor to high, but I still don't know how to get the I2C communication working, as there is nothing about it in docs. 

THANKS

**Comment by fdufnews** [comment]
See here [https://github.com/fdufnews/pico_examples](https://github.com/fdufnews/pico_examples) some code I collected here and there and some I wrote myself while I started using the Pico.
There is a _driver_ directory where you'll find a driver for de SSD1306 and in the _examples_ directory there is a 1306oled one with some example code. Maybe you'll have to modify the pinout if you are not using the default one for the I2C.
Concerning the doc, here you'll find all you can expect.
https://docs.micropython.org/en/latest/
Concerning I2C it is here
https://docs.micropython.org/en/latest/library/machine.I2C.html

**Reply by mlajciak** [reply]
Hi, I've tested the i2c_1306oled.py, and got errir that says that there are no I2C deveices connected, however I've checked the circuit with multimeter and it's wired correctly. Also when I connect for example mpu6050 gyro to i2c it works so I2C is not damaged i think that your code is not going to work with the waveshare 8 pin IO display.

**Reply by fdufnews** [reply]
Maybe the pullup resistors on SCK and SDA are missing as this board is intended first to be used with the SPI bus.

Edit: I just found the schematic on Waveshare's wiki, I confirm there are no pullup resistors on this board, so you have to add both.

**Reply by mlajciak** [reply]
I tried that, i added 2k resistor to both sda and scl to 5v, but it's still not finding  device.

**Comment by fdufnews** [comment]
> i added 2k resistor to both sda and scl to 5v

The pullups shall be connected to 3.3V not 5V. Both circuits work on 3.3V.
<hr>
Have you seen this on the waveshare's wiki?

> When using I2C: The high and low **state of DC** can **control the address of the slave device**. If it is connected to GND, then the 7-bit address of I2C is: 0x3C;

If left floatting the address can be unstable
<hr>
There is a comment on the SSD1306 datasheet concerning the switch to I2C. On the schematic there is a 0Ω resistor (at the left of BS0 BS1) that seems to connect the 2 I/Os but maybe you can check that it is at the expected place.

![image](https://github.com/micropython/micropython/assets/4077715/2066be0d-0ff6-46e8-8a47-eb0e995915ef)

**Comment by peterhinch** [comment]
The SSD1306 (in I2C or SPI mode) is also supported by [nano-gui](https://github.com/peterhinch/micropython-nano-gui/tree/master) and [micro-gui](https://github.com/peterhinch/micropython-micro-gui). The driver is [here](https://github.com/peterhinch/micropython-nano-gui/blob/master/drivers/ssd1306/ssd1306.py).
